pkg/tracing/config.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tracing provides the canonical tracing functionality used by Go-based
+// Istio components.
+//
+// The package provides direct integration with the Cobra command-line processor which makes it
+// easy to build programs that use a consistent interface for tracing. Here's an example
+// of a simple Cobra-based program using this tracing package:
+//
+//		func main() {
+//			// get the default tracing options
+//			options := tracing.NewOptions()
+//
+//			rootCmd := &cobra.Command{
+//				Run: func(cmd *cobra.Command, args []string) {
+//
+//					// configure the tracing system
+//					if _, err := tracing.Configure(options); err != nil {
+//                      // print an error and quit
+//                  }
+//
+//					// Generate a trace
+//					ot.StartSpan(...)
+//				},
+//			}
+//
+//			// add tracing-specific flags to the cobra command
+//			options.AttachCobraFlags(rootCmd)
+//			rootCmd.SetArgs(os.Args[1:])
+//			rootCmd.Execute()
+//		}
+//
+// The point of this package is to configure the global OpenTracing tracer
+// for the process as read from ot.GlobalTracer and used by ot.StartSpan.
+package tracing
+
+import (
+	"fmt"
+	"io"
+	"time"
+
+	ot "github.com/opentracing/opentracing-go"
+	jaeger "github.com/uber/jaeger-client-go"
+	"github.com/uber/jaeger-client-go/transport"
+	"github.com/uber/jaeger-client-go/transport/zipkin"
+	"go.uber.org/zap"
+
+	"istio.io/istio/pkg/log"
+)
+
+type holder struct {
+	closer io.Closer
+	tracer ot.Tracer
+}
+
+var (
+	httpTimeout = 5 * time.Second
+	sampler     = jaeger.NewConstSampler(true)
```

Comment:
we probably want to think about:
- only tracing when trace context information is already present (mixer)
- for some percentage of requests (pilot)

Reply:
You're suggesting adding these as TODO items in the code I presume?

Reply:
Yes. TODOs. This is more thinking out loud. I'm a bit concerned that the high number of pilot requests (1 per sec per envoy ?) may generate a TON of traces.